scripts/eqtl_hail_batch/generate_eqtl_spearman.py

```diff
@@ -0,0 +1,448 @@
+    # locus, and alleles
+    t = hl.Table.from_pandas(spearman_df)
+    t = t.annotate(global_bp=hl.locus(t.chrom, hl.int32(t.bp)).global_position())
+    t = t.annotate(locus=hl.locus(t.chrom, hl.int32(t.bp)))
+    # get alleles
+    mt = mt.key_rows_by('locus')
+    t = t.key_by('locus')
+    t = t.annotate(
+        a1=mt.rows()[t.locus].alleles[0],
+        a2=hl.if_else(
+            hl.len(mt.rows()[t.locus].alleles) == 2, mt.rows()[t.locus].alleles[1], 'NA'
+        ),
+    )
+    # add ID annotation after adding in alleles, a1, and a2
+    t = t.annotate(
+        id=hl.str(':').join(
+            [
+                hl.str(t.chrom),
+                hl.str(t.bp),
+                t.a1,
+                t.a2,
+                t.gene_symbol,
+                # result.db_key, # cell_type_id (eg nk, mononc)
+                hl.str(t.round),
+            ]
+        )
+    )
+    print(f'Printing final table: {t.show()}')
+    spearman_df = t.to_pandas()
+    return spearman_df
+
+
+def merge_df_and_convert_to_string(*df_list):
+    """Merge all Spearman dfs and convert to string using .to_string() on df"""
+    merged_df: pd.DataFrame = pd.concat(df_list)
+    pvalues = merged_df['p_value']
+    fdr_values = pd.DataFrame(list(multi.fdrcorrection(pvalues))).iloc[1]
+    merged_df = merged_df.assign(fdr=fdr_values)
+    merged_df['fdr'] = merged_df.fdr.astype(float)
+    return merged_df.to_string()
+
+
+# Create click command line to enter dependency files
+@click.command()
+@click.option(
+    '--expression', required=True, help='A sample x gene TSV of expression values'
+)
+@click.option(
+    '--geneloc', required=True, help='A TSV of start and end positions for each gene'
+)
+@click.option(
+    '--covariates', required=True, help='A TSV of covariates to calculate residuals'
+)
+@click.option(
+    '--keys',
+    required=True,
+    help='A TSV of sample ids to convert external to internal IDs',
+)  # pylint: disable=too-many-locals
+@click.option(
+    '--output-prefix',
+    required=True,
+    help='A path prefix of where to output files, eg: gs://MyBucket/output-folder/',
+)
+def main(
+    expression: str,
+    geneloc,
+    covariates,
+    keys,
+    output_prefix: str,
+):
+    """
+    Creates a Hail Batch pipeline for calculating EQTLs
+    """
+    dataset = os.getenv('CPG_DATASET')
+    backend = hb.ServiceBackend(billing_project=dataset, remote_tmpdir=remote_tmpdir())
+    batch = hb.Batch(name='eQTL', backend=backend, default_python_image=DRIVER_IMAGE)
+
+    # load in files to get number of scatters and residuals
+    expression_df_literal = pd.read_csv(AnyPath(expression), sep='\t')
+    geneloc_df_literal = pd.read_csv(AnyPath(geneloc), sep='\t')
+    covariate_df_literal = pd.read_csv(AnyPath(covariates), sep=',')
+
+    # only run batch jobs if there are test genes in the chromosome
+    n_genes_in_scatter = get_number_of_scatters(expression_df_literal, geneloc_df_literal)
+    if n_genes_in_scatter > 0:
```

Comment:
[nitpicking]
In what situations it can equal 0? Wonder it makes sense to print some error message to explicitly catch invalid inputs.

Also, in terms of code style, it's preferable to avoid unnecessary indentation if possible, so instead of

```
if n_genes_in_scatter > 0:
    <a lot of code>
```

Better do:

```
if n_genes_in_scatter <= 0:
    raise ValueError('')

<a lot of code>
```

Reply:
I've added in the `raise ValueError('')` message here and fixed the indentation. Good to know about the code style, I didn't know about that